pkg/apis/resource/types.go

```diff
@@ -463,10 +469,97 @@ type DeviceRequest struct {
 	// +optional
 	// +featureGate=DRAAdminAccess
 	AdminAccess *bool
+
+	// FirstAvailableOf contains subrequests, exactly one of which must be satisfied
+	// in order to satisfy this request. This field may only be set in the
+	// entries of DeviceClaim.Requests.
+	//
+	// +optional
+	// +oneOf=deviceRequestType
+	// +listType=atomic
+	// +featureGate=DRAPrioritizedList
+	FirstAvailableOf []DeviceSubRequest
```

Comment:
> This is not the behavior that the scheduler in 1.32 implements. We cannot retro-actively make that behavior a requirement.

Fair. So, let's continue to allow empty claims. 

> That user experience then would be worse because pods could get scheduled without the devices requested via RankedRequests.

This is where I disagree. I am looking at the integral of user pain over the remaining lifetime of K8s. I don't think optimizing for "the best UX" for a tiny portion of users during short transitional phases should take precedence over the long term UX for the all users of the API. This is why I think the rules should be 1) Almost inviolably, things that worked before should continue to work; 2) Best effort, new fields should cause old code to reject gracefullly.

It's 2) that's in question here - I don't think that should be inviolable, just best effort. I don't think making a worse long-term API so that old schedulers will reject things they don't understand is necessary. My proposal would be this:

* Continue to allow empty claims because not doing that would violate 1)
* Allow the additional of `RankedRequests` next to `Requests`. Old schedulers will not allocate anything from `RankedRequests`. This means during the transition period, an old scheduler may allocate a claim it did not fully understand, rather than refusing to allocate it. I think that's an acceptable compromise to get a better long term UX.
* Correcting my previous snippet, RankedRequest should be a new struct, consisting of a name and a slice of `DeviceSubRequest` values (so you can have multiple ranked requests that each resolve to a single satisfied request).

Reply:
> 1) Almost inviolably, things that worked before should continue to work; 2) Best effort, new fields should cause old code to reject gracefullly.

Agreed.

> I don't think making a worse long-term API 

So the argument is that `RankedRequests` is a better API? That wasn't clear to me from [what you wrote](https://github.com/kubernetes/kubernetes/pull/128586#pullrequestreview-2493616997) and it's not obviously better to me. Care to explain?

Reply:
Sure. Right now, we have a struct where many fields aren't used, when you set the `FirstAvailableOf`, which in that case nests a very similar struct. This is a bit awkward, it looks a bit like Service which has this same pattern of sets of fields that are present or not, depending on the value of another field; this pattern has been challenging over the years.

It's a matter of opinion/taste though. Expressiveness is the same for both. So, ultimately I am OK with either approach.

Reply:
Agreed, OpenAPI definition and client-side validation is worse when extending DeviceRequest. But I am not sure how much it matters. I suspect it could also confuse users if they can't put `firstAvailableOf` into `requests`. The YAML probably looks a bit better when using `requests` - IMHO.